agent/cache/cache.go

```diff
@@ -609,13 +604,93 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 	// identical calls to fetch will return the same waiter rather than
 	// perform multiple fetches.
 	entry.Fetching = true
+	entry.HasGoroutine = true
 	c.entries[key] = entry
 	metrics.SetGauge([]string{"consul", "cache", "entries_count"}, float32(len(c.entries)))
 	metrics.SetGauge([]string{"cache", "entries_count"}, float32(len(c.entries)))
 
-	tEntry := r.TypeEntry
 	// The actual Fetch must be performed in a goroutine.
-	go func() {
+	go c.launchBackgroundFetcher(key, r, entry)
```

Comment:
We mark that the entry `HasGoroutine=true` and then launch the looping goroutine. It has a defer to flip that back to `false` so it should allow for a subsequent call to re-launch the goroutine should it become needed again.